src/server/snapshot.cc

```diff
@@ -95,246 +94,236 @@ void SliceSnapshot::Join() {
+}
 
-    io::Result<uint8_t> res = tmp_serializer.SaveEntry(pkey, *entry.pval_ptr, entry.expire_ms);
-    CHECK(res);  // we write to StringFile.
+void SliceSnapshot::CloseRecordChannel() {
+  // stupid barrier to make sure that SerializePhysicalBucket finished.
+  // Can not think of anything more elegant.
+  mu_.lock();
+  mu_.unlock();
 
-    io::StringFile sfile;
-    error_code ec = tmp_serializer.FlushToSink(&sfile);
-    CHECK(!ec && !sfile.val.empty());
-    PushFileToChannel(&sfile, entry.db_ind, 1, false);
+  // Make sure we close the channel only once with a CAS check.
+  bool expected = false;
+  if (closed_chan_.compare_exchange_strong(expected, true)) {
+    dest_->StartClosing();
   }
 }
 
-unsigned SliceSnapshot::SerializePhysicalBucket(DbIndex db_index, PrimeTable::bucket_iterator it) {
-  DCHECK_LT(it.GetVersion(), snapshot_version_);
-
-  // traverse physical bucket and write it into string file.
-  it.SetVersion(snapshot_version_);
-  unsigned result = 0;
-
-  lock_guard lk(mu_);
+SliceSnapshot::DbRecord SliceSnapshot::GetDbRecord(DbIndex db_index, std::string value) {
+  auto id = rec_id_++;
+  DVLOG(2) << "Pushed " << id;
 
-  if (db_index == savecb_current_db_) {
-    while (!it.is_done()) {
-      ++result;
-      SerializeSingleEntry(db_index, it->first, it->second, rdb_serializer_.get());
-      ++it;
-    }
-    num_records_in_blob_ += result;
-  } else {
-    bool serializer_compression = (compression_mode_ != CompressionMode::NONE);
-    RdbSerializer tmp_serializer(serializer_compression);
+  stats_.channel_bytes += value.size();
+  return DbRecord{.db_index = db_index, .id = id, .value = std::move(value)};
+}
 
-    while (!it.is_done()) {
-      ++result;
-      SerializeSingleEntry(db_index, it->first, it->second, &tmp_serializer);
-      ++it;
-    }
-    io::StringFile sfile;
-    error_code ec = tmp_serializer.FlushToSink(&sfile);
-    CHECK(!ec && !sfile.val.empty());
-    PushFileToChannel(&sfile, db_index, result, false);
+SliceSnapshot::BorrowedSerializer::BorrowedSerializer(SliceSnapshot* snapshot, DbIndex index,
+                                                      CompressionMode compression_mode)
+    : db_index{index}, snapshot{snapshot}, serializer{nullopt} {
+  if (index != snapshot->current_db_) {
+    serializer.emplace(compression_mode != CompressionMode::NONE);
   }
-  return result;
 }
 
-void SliceSnapshot::PushFileToChannel(io::StringFile* sfile, DbIndex db_index, unsigned num_records,
-                                      bool should_compress) {
-  string string_to_push = std::move(sfile->val);
-
-  if (should_compress) {
-    if (!zstd_serializer_) {
-      zstd_serializer_.reset(new ZstdCompressSerializer());
-    }
-    auto comp_res = zstd_serializer_->Compress(string_to_push);
-    if (comp_res.first) {
-      string_to_push.swap(comp_res.second);
-    }
+SliceSnapshot::BorrowedSerializer::~BorrowedSerializer() {
+  if (serializer) {
```

Comment:
Although I understand that this flow working correctly I find this flow a bit confusing. 
To understand the flow you need to think when this push to channel is happening on the destructor, and also I need to see when the destructor is called to verify correctness, its not so intuitive .
I would like to suggest a different flow:
SliceSnapshot will hold serializer per db index i.e vector of serializers, SerializeEntry will call the relevant serializer in the vector.
In the function FlushDefaultBuffer we will iterate over the serializer and push all to channel.
Another value to this suggestion is if the data in the flow of OnDBChange is small today we have to flush, but if we flush it just when its bigger than some size we can do better compression and less records pushed to channel.
I am not so familiar with the flow of OnJournalEntry let me know if this change makes sense here as well
and in general if you think suggestion make sense to you.

Reply:
Keeping a list of serializers would be over-engineering because we need to strictly keep the order, so we'd need to flush after any interrupting operation either way.

I can change it back if its confusing - this will be the simplest solution. The approach might be not very intuitive, but its a more or less typical RAII